chirp-impl/src/main/java/sample/chirper/chirp/impl/ChirpServiceImpl.java

```diff
@@ -3,162 +3,80 @@
+        this.chirps = chirps;
+        this.db = db;
+
+        persistentEntitites.register(ChirpTimelineEntity.class);
+    }
 
-        // We currently ignore the fact that it is possible to get duplicate chirps
-        // from the recent and the topic. That can be solved with a de-duplication stage.
-        return Source.from(recentChirps).concat(publishedChirps);
-      });
-    };
-  }
+    @Override
+    public ServiceCall<Chirp, NotUsed> addChirp(String userId) {
+        return chirp -> {
+            if (!userId.equals(chirp.userId))
+                throw new IllegalArgumentException("UserId " + userId + " did not match userId in " + chirp);
 
-  @Override
-  public ServiceCall<HistoricalChirpsRequest, Source<Chirp, ?>> getHistoricalChirps() {
-    return req -> {
-      List<Source<Chirp, ?>> sources = new ArrayList<>();
-      for (String userId : req.userIds) {
-          Source<Chirp, NotUsed> select = db
-            .select("SELECT * FROM chirp WHERE userId = ? AND timestamp >= ? ORDER BY timestamp ASC", userId,
-                req.fromTime.toEpochMilli())
-            .map(this::mapChirp);
-        sources.add(select);
-      }
-        // Chirps from one user are ordered by timestamp, but chirps from different
-        // users are not ordered. That can be improved by implementing a smarter
-        // merge that takes the timestamps into account.
-      Source<Chirp, ?> result = Source.from(sources).flatMapMerge(sources.size(), s -> s);
-      return CompletableFuture.completedFuture(result);
-    };
-  }
+            return persistentEntitites.refFor(ChirpTimelineEntity.class, userId)
+                    .ask(new AddChirp(chirp))
+                    .thenApply(done -> NotUsed.getInstance());
+        };
+    }
 
-  private Chirp mapChirp(Row row) {
-    return new Chirp(row.getString("userId"), row.getString("message"), 
-        Optional.of(Instant.ofEpochMilli(row.getLong("timestamp"))), Optional.of(row.getString("uuid")));
-  }
 
-  private CompletionStage<PSequence<Chirp>> recentChirps(PSequence<String> userIds) {
-    int limit = 10;
-    PSequence<CompletionStage<PSequence<Chirp>>> results = TreePVector.empty();
-    for (String userId : userIds) {
-      CompletionStage<PSequence<Chirp>> result = db
-          .selectAll("SELECT * FROM chirp WHERE userId = ? ORDER BY timestamp DESC LIMIT ?", userId, limit)
-          .thenApply(rows -> {
-            List<Chirp> chirps = rows.stream().map(this::mapChirp).collect(Collectors.toList());
-            return TreePVector.from(chirps);
-          });
-      results = results.plus(result);
+    @Override
+    public ServiceCall<LiveChirpsRequest, Source<Chirp, ?>> getLiveChirps() {
+        return req -> chirps.getRecentChirps(req.userIds).thenApply(recentChirps -> {
+            List<Source<Chirp, ?>> sources = new ArrayList<>();
+            for (String userId : req.userIds) {
+                sources.add(topic.subscriber(userId));
+            }
```

Comment:
Yes, I suppose that's true. Right now it relies on the fact that no queries with duplicate IDs will ever be made.

If you think it's worth de-duplicating on the back end, let's do that in a separate PR.